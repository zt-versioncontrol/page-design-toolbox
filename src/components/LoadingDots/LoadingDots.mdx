import { Meta, ArgTypes } from '@storybook/blocks'
import LoadingDots from '@/components/LoadingDots/LoadingDots.vue'
import * as Stories from './LoadingDots.stories.ts'
import { ComponentSummary, Examples, ExampleGroup } from '/.storybook/documents/DocBlocks'

<Meta title="Loading Dots Document" of={Stories} component={LoadingDots} />

<ComponentSummary of={Stories} />

<ArgTypes of={Stories} />

<Examples>
  <ExampleGroup stories={[Stories.Default]} title={'Default size'} />
  <ExampleGroup stories={[Stories.Small, Stories.Large]} title={'CustomSize'} />
  <ExampleGroup title={'Colored Dots'} stories={[Stories.GreenDots, Stories.RedDots]} />
</Examples>
