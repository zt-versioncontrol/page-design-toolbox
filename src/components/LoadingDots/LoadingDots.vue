<script setup lang="ts">
import { MAIN_TEXT } from '@/standards/Colors'
import { LOADING_ANIMATION_CYCLE } from '@/standards/Time'
import { INTEGRATED_ICON } from '@/standards/Dimensions'

export interface Props {
  /**
   * Dots size. May be and CSS size value
   */
  size?: string
  /**
   * Dots color. Maybe any CSS color
   */
  color: string
}

// TODO: use standard dimensions

const props = withDefaults(defineProps<Props>(), {
  size: INTEGRATED_ICON,
  color: MAIN_TEXT
})
</script>

<template>
  <div class="dots-container">
    <div class="loading-dot dot-first" data-test="first-dot"></div>
    <div class="loading-dot dot-second" data-test="second-dot"></div>
    <div class="loading-dot dot-third" data-test="third-dot"></div>
  </div>
</template>

<style scoped>
@keyframes flashing {
  0% {
    opacity: 0.5;
    transform: scale(0.8);
  }
  50% {
    opacity: 1;
    transform: scale(1);
  }
  100% {
    opacity: 0.5;
    transform: scale(0.8);
  }
}

.dots-container {
  display: flex;
  gap: 0.5%;
}

.loading-dot {
  height: v-bind(size);
  width: v-bind(size);
  border-radius: 200%;
  background-color: v-bind(color);
}

.dot-first {
  animation-name: flashing;
  animation-duration: v-bind('LOADING_ANIMATION_CYCLE + "s"');
  animation-iteration-count: infinite;
  animation-delay: 0s;
}

.dot-second {
  animation-name: flashing;
  animation-duration: v-bind('LOADING_ANIMATION_CYCLE + "s"');
  animation-iteration-count: infinite;
  /* To synchronize with first and third dots. it should start animation midway between them */
  animation-delay: v-bind('LOADING_ANIMATION_CYCLE / 4 + "s"');
}

.dot-third {
  animation-name: flashing;
  animation-duration: v-bind('LOADING_ANIMATION_CYCLE + "s"');
  animation-iteration-count: infinite;
  /* To synchronize with first dot, third dot should start animation when first dot is halfway through */
  animation-delay: v-bind('LOADING_ANIMATION_CYCLE / 2 + "s"');
}
</style>
