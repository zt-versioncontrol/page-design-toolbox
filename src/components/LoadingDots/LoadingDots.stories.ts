import type { Meta, StoryObj } from '@storybook/vue3'
import LoadingDots from '@/components/LoadingDots/LoadingDots.vue'
import { type Documentation } from '@/../.storybook/documents/DocUtilities'
import * as componentTags from '@/../.storybook/documents/ComponentTags'
import * as componentBadges from '@/../.storybook/documents/ComponentBadges'

const meta: Meta<typeof LoadingDots> = {
  title: 'Interactivity/Loadingdots',
  component: LoadingDots,
  excludeStories: ['documentation']
}

export default meta

type Story = StoryObj<typeof LoadingDots>

export const documentation: Documentation = {
  title: 'Loading Dots',
  summary:
    'Loading dots is used to indicate pending operations. Mainly used while waiting for incoming text',
  tags: [componentTags.ICON],
  badges: [componentBadges.RELEASED]
  // TODO: add testing guide
}

export const Default: Story = {
  args: {}
}

export const Small: Story = {
  args: {
    size: '0.2em'
  }
}

export const Large: Story = {
  args: {
    size: '5em'
  }
}

export const GreenDots: Story = {
  args: {
    color: 'green'
  }
}

export const RedDots: Story = {
  args: {
    color: 'red'
  }
}
