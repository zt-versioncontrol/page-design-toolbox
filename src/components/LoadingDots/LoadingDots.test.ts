import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import LoadingDots from '@/components/LoadingDots/LoadingDots.vue'

describe('Default component', () => {
  it('should be rendered', () => {
    const wrapper = mount(LoadingDots, {
      global: {
        plugins: []
      },
      attachTo: document.body
    })

    expect(wrapper.isVisible()).toBe(true)
  })

  it('Match Snapshot', () => {
    const wrapper = mount(LoadingDots, {
      global: {
        plugins: []
      }
    })

    expect(wrapper.element).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })
})
