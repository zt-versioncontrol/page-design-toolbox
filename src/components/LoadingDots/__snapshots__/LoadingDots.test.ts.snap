// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Default component > Match Snapshot 1`] = `
<div
  class="dots-container"
  data-v-cb975c38=""
>
  <div
    class="loading-dot dot-first"
    data-test="first-dot"
    data-v-cb975c38=""
  />
  <div
    class="loading-dot dot-second"
    data-test="second-dot"
    data-v-cb975c38=""
  />
  <div
    class="loading-dot dot-third"
    data-test="third-dot"
    data-v-cb975c38=""
  />
</div>
`;
